Email editor - button and buttons blocks [MAILPOET-5644]

costasovo opened the pull request:
## Description
This PR adds support for buttons and button blocks to the email editor.

## Code review notes
- colors set via CSS class don't work in renderer (we have an extra ticket for that 5741)

## QA notes
I would suggest skipping QA for now and doing QA after we finish bug fixes and refactoring the renderer.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-5644]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5644]: https://mailpoet.atlassian.net/browse/MAILPOET-5644?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo committed:
Enable blocks and register dummy button/buttons renderers

[MAILPOET-5644]

costasovo committed:
Enable border-related settings in the email editor

[MAILPOET-5644]

costasovo committed:
Fix block width preprocessor for blocks with percentage width

[MAILPOET-5644]

costasovo committed:
Add renderer for core/button block

[MAILPOET-5644]

costasovo committed:
Disable styles selection for button block

As for now, we can't support preset styles for buttons as they
don't update block attributes we read in renderer
[MAILPOET-5644]

costasovo committed:
Remove unnecessary eslint ignore from image and column hooks

[MAILPOET-5644]

costasovo committed:
Add reduced flex layout for buttons block

[MAILPOET-5644]

costasovo committed:
Disallow button block being inserted directly to a column

[MAILPOET-5644]

costasovo committed:
Add default core editor styles

We accidentally removed them when adding layout styles.
[MAILPOET-5644]

costasovo committed:
Add block toolbar controls for flex email layout

[MAILPOET-5644]

costasovo committed:
Add width and height to the image, select editor modal

This fixes the issue that the modal resizes after the image is loaded.
[MAILPOET-5644]

costasovo committed:
Render buttons with borders with the correct width

[MAILPOET-5644]

costasovo committed:
Add buttons block rendering

[MAILPOET-5644]

costasovo committed:
Improve font styles rendering in button

[MAILPOET-5644]

costasovo committed:
Remove forgotten bdumps from image block

[MAILPOET-5644]

costasovo committed:
Split button and buttons hooks into two files

[MAILPOET-5644]

costasovo committed:
Override default button border-radius

The default button radius is usually defined in theme.json.
The default from the core is 9999px, which is not convenient for email
renderer.

[MAILPOET-5644]

costasovo committed:
Add support for text decoration to typography preprocessor

costasovo committed:
Apply text-decoration in the button renderer from preprocessor

[MAILPOET-5644]

costasovo committed:
Extract flex layout rendering functionality to an extra class

[MAILPOET-5644]

costasovo committed:
Add Flex Renderer test

[MAILPOET-5644]

costasovo committed:
Remove witdh:100% from columns wrapper

This was causing incorrect width in cases when $layoutPaddingLeft
and $layoutPaddingRight were set. Box-sizing doesn't have proper
support in clients, so I tried to fix it by removing 100% of the width.
According to tests in AcidEmail, this shouldn't break anything.

This was introduced when we added full-width alignment.
[MAILPOET-5644]

costasovo committed:
Skip rendering of empty buttons

[MAILPOET-5644]

costasovo committed:
Unify styles processing in button renderer with other blocks

[MAILPOET-5644]

costasovo committed:
Add Button renderer test

This commit adds a button renderer test and fixes a couple of issues
found when working on the test
[MAILPOET-5644]

costasovo committed:
Add schema to email editor theme.json

[MAILPOET-5644]

costasovo committed:
Add base text and background color to theme.json

[MAILPOET-5644]

costasovo committed:
Use theme.json defined base colors in editor

[MAILPOET-5644]

costasovo committed:
Use theme.json defined background in renderer

[MAILPOET-5644]

costasovo committed:
Improve typography preprocessor to distribute font color from theme.json

[MAILPOET-5644]

costasovo committed:
Apply font color from the preprocessor in the button block

[MAILPOET-5644]

lysyjan commented on mailpoet/lib/EmailEditor/Engine/Renderer/Preprocessors/BlocksWidthPreprocessor.php:
```diff
@@ -23,7 +23,9 @@ public function preprocess(array $parsedBlocks, array $layoutStyles): array {
 
-      $width = $this->convertWidthToPixels($block['attrs']['width'] ?? '100%', $layoutWidth);
+      $widthInput = $block['attrs']['width'] ?? '100%';
+      $widthInput = is_numeric($widthInput) ? "$widthInput%" : $widthInput;
```
I guess I understand that this is a necessary check, but maybe a comment about the reasons would be more transparent for other developers. If I am not mistaken, the reason is the different width values of some blocks, such as buttons.
What do you think about a comment here?

costasovo replied:
💯 I was thinking about adding a comment there but forgot. Thanks for the heads up!

lysyjan commented on mailpoet/lib/EmailEditor/Integrations/Core/Renderer/Blocks/Button.php:
```diff
@@ -28,7 +28,13 @@ public function render($blockContent, array $parsedBlock, SettingsController $se
+    // Parent block prepares container with proper width. If the width is set let's use full width of the container
+    // otherwise let's use auto width.
+    $width = 'auto';
+    if (($parsedBlock['attrs']['width'] ?? null)) {
```
Using the null coalescing operator in this condition doesn't make sense to me. I would say `isset` could work better here. But maybe I'm missing something. What do you think about it? There are more similar cases.

costasovo replied:
Oh yes. I'll use isset. I don't know why I used the operator.

lysyjan replied:
I would say gradual code modifications and looking for a working solution caused this.

lysyjan commented on mailpoet/tests/integration/EmailEditor/Integration/Core/Renderer/Blocks/ButtonTest.php:
```diff
@@ -0,0 +1,164 @@
+      'topRight' => '2px',
+      'bottomLeft' => '3px',
+      'bottomRight' => '4px',
+      ];
```
This is just a small remark because the indent is incorrect here.
Also, I have noticed that all methods don't have a return type. Personally, I try to use void in tests, but maybe it's only a personal preference.

lysyjan reviewed (commented):
A huge piece of work! Thanks for many fixes and improvements 👍 

During my brief testing, I noticed that the browser preview contains an unexpected gap when there are other blocks around.

Editor:
![Screenshot 2024-01-04 at 15 50 26](https://github.com/mailpoet/mailpoet/assets/13644846/8349019c-8673-46d6-97cd-5b81f4937628)

Preview:
![Screenshot 2024-01-04 at 15 50 50](https://github.com/mailpoet/mailpoet/assets/13644846/0654d1f2-4856-4fc1-898c-497111092bbb)

I would say it's caused by setting font-size on a different HTML element than is needed.

lysyjan commented:
@costasovo I almost forgot about another idea here. When I was trying to use the button block, a missing feature was the vertical spacing configuration. Have you considered something like that? The post editor allows setting vertical margins. I know about a possible workaround when I wrap buttons by columns, but something like that would be more straightforward.

costasovo committed:
Use isset instead ?? operator in button renderer

[MAILPOET-5644]

costasovo committed:
Pass default font-size in email attrs in TypographyPreprocessor

It is convenient to always know font-size in nested blocks so that we don't
have to read it from the settings controller.
[MAILPOET-5644]

costasovo committed:
Render font size on button instead of buttons wrapper

Font size on the wrapper was causing whitespace below the buttons block.
[MAILPOET-5644]

costasovo committed:
Fix formatting and add return type hints to buttons related tests

[MAILPOET-5644]

costasovo committed:
Add comment about numeric values in BlocksWidthPreprocessor

[MAILPOET-5644]

costasovo commented:
@lysyjan Thank you for the review! I addressed all your suggestions in the new five commits.

lysyjan reviewed (approved):
Thank you! Great job
